src/SFA.DAS.EAS.Portal/Application/Commands/Cohort/CohortApprovalRequestedCommandHandler.cs

```diff
@@ -0,0 +1,62 @@
+﻿using SFA.DAS.Commitments.Api.Client.Interfaces;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Linq;
+using System;
+using SFA.DAS.EAS.Portal.Application.Services;
+using SFA.DAS.EAS.Portal.Types;
+
+namespace SFA.DAS.EAS.Portal.Application.Commands.Cohort
+{
+    public class CohortApprovalRequestedCommandHandler : ICommandHandler<CohortApprovalRequestedCommand>
+    {
+        private readonly IAccountDocumentService _accountsService;
+        private readonly IProviderCommitmentsApi _providerCommitmentsApi;
+
+        public CohortApprovalRequestedCommandHandler(
+            IAccountDocumentService accountsService, 
+            IProviderCommitmentsApi providerCommitmentsApi)
+        {
+            _accountsService = accountsService;
+            _providerCommitmentsApi = providerCommitmentsApi;
+        }
+
+        public async Task Handle(CohortApprovalRequestedCommand command, CancellationToken cancellationToken = default)
```

Comment:
This handler assumes that the account document has already been created, and will throw a `NullReferenceException` if that assumption is incorrect.

That means that if we haven't received and successfully processed an event that can create the account, before the last retry of handling the `CohortApprovalRequestedByProvider` event, the event will end up in the error queue, and it will require a manual operation to replay the event.

We could potentially create the account in this handler, if it doesn't currently exist.

We need to decide on how we handle out-of-order events from the different source systems. I think we should research current best practices, and possibly call on the architects for guidance, as this is quite central to how the portal operates.

Reply:
One opinion found on t'internet (I'm sure an opposing opinion can be found 😄), from [here](https://www.ben-morris.com/data-design-for-event-driven-architecture-autonomy-encapsulation-and-ordering/)... (this is also the way I'm leaning - handle events out of order, i.e. create the account if it doesn't already exist, even if logically it already should)

> **Ordering shouldn’t matter**
> Another common mistake is to confuse events with feeds. An event should be a discrete item that does not rely on previous events to describe current state.
> 
> Given that events are time-based it can be logical to assume that they should be processed in a particular order. The problem here is that networks are inherently unreliable so it is very difficult to guarantee the order in which events are delivered. It is even more difficult to guarantee the order in which they are processed.
> 
> The easiest way of solving the ordering problem is to design events so the order in which they are processed does not matter. Bear in mind that if you rely on message ordering then you are effectively coupling your applications together in a temporal, or time-based, sense.